src/app/main.rs

```diff
@@ -345,13 +345,19 @@ where
                 // A per-`dst::Route` layer that uses profile data to configure
                 // a per-route layer.
                 //
-                // The `classify` module installs a `classify::Response`
-                // extension into each request so that all lower metrics
-                // implementations can use the route-specific configuration.
+                // 1. The `classify` module installs a `classify::Response`
+                //    extension into each request so that all lower metrics
+                //    implementations can use the route-specific configuration.
+                // 2. A timeout is optionally enabled if the target `dst::Route`
+                //    specifies a timeout. This goes before `retry` to cap
+                //    retries.
+                // 3. Retries are optionally enabled depending on if the route
+                //    is retryable.
                 let dst_route_layer = phantom_data::layer()
                     .push(insert_target::layer())
                     .push(metrics::layer::<_, classify::Response>(retry_http_metrics.clone()))
                     .push(retry::layer(retry_http_metrics))
+                    .push(proxy::http::timeout::layer())
```

Comment:
Looking at the metrics output, since this layer resolves the error into an HTTP response directly, it gets counted in the `route_http_metrics` as normal response with `status_code="504"`. I wonder if that is misleading, suggesting that the other service actually returned a 504. If so, I can put this *after* the `route_http_metrics` layer.

Reply:
I haven't thought about this exhaustively, but I think that there isn't any other good way to determine that this proxy returned the 5xx (as opposed to another). In that case, we probably need to reorder this with the metrics layer, and verify that the metrics layer has a reasonable way to format the error in an unambiguous way.

As to whether this should be a label or a new metric...

On one hand, it's convention to use `response_total - request_total` to indicate the number of in-flight requests. So this would suggest using a label.

I suspect this should fit into the metrics classify api, and the classifier type would have to be aware of the types of errors it can receive so that it can format them appropriately?

Reply:
I would expect `route_actual_response_total` to reflect the actual responses from the service and `route_response_total` to reflect the responses that the proxy sends back to the client.  Therefore, in the case where the request to the service times out, I would expect `route_actual_response_total` to be incremented with a label like `{error=timeout}` or something like that and `route_response_total` to be incremented with `status_code=504`.  

Does that make sense?  We can tell that the service itself never returned a 504 because `route_actual_response_total{status_code=504}` will be zero.

Reply:
Latest commit adds this to the metrics (text copied straight from the unit test):

```
route_response_total{direction="outbound",dst="profiles.test.svc.cluster.local:80",status_code="504",classification="failure",error="timeout"} 1
```